### PR TITLE
Refactor lua loader

### PR DESCRIPTION
**Pull request by contrun**
This PR removes a gigantic macro, adds a function to get intermediate result and renames a few functions. There is no functionality change. It just makes future life easier.

### CODE REVIEW

**Comment by XuJiandong** [comment]
Merge changes from the master branch and the CI will work.

File: lua-loader/lua-ckb.c [hunk 1]
```diff
@@ -51,56 +66,90 @@ typedef struct {
...
+        return CKB_LUA_OUT_OF_MEMORY;
+    }
+
+    ret = call_syscall(f, buf);
+    if (ret != 0) {
+        free(buf);
+        return ret;
+    }
+
+    /* We have allocated a buffer with a size larger than what is needed */
```
Comment (XuJiandong) [inline lua-loader/lua-ckb.c:123]: Why not just malloc the smaller size one? The realloc operation is not necessary.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

**Comment by XuJiandong** [comment]
Add a test case: loop 10000 times; Every time, it reads 100K data. Then watch if it has a memory leak issue or not.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: lua-loader/lua-ckb.c [hunk 2]
```diff
@@ -51,56 +66,90 @@ typedef struct {
...
+    }
+    PANIC("invalid number of extra arguments %d", f->num_extra_arguments);
+    return -1;
+}
+
+int call_syscall_get_result(BUFFER_T *result, struct syscall_function_t *f) {
+    int ret = 0;
+    /* Only obtain the minimal buffer length required */
+    if (f->length != NULL && *f->length == 0) {
+        ret = call_syscall(f, NULL);
```
Comment (XuJiandong) [inline lua-loader/lua-ckb.c:94]: When the length is fetched, the `offset` should be passed in. Different `offset` get different lengths.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: lua-loader/lua-ckb.c (hunk 1 above)
Comment (contrun) [reply lua-loader/lua-ckb.c:123]: The realloc is only necessary when the lua script passed a length argument and the actual data size is less than that length. realloc only shrink the memory space size.

Comment (contrun) [reply lua-loader/lua-ckb.c:123]: I think if the lua script explicitly passed a length, we should alloc that number of memory.

File: lua-loader/lua-ckb.c [hunk 3]
```diff
@@ -51,56 +66,90 @@ typedef struct {
...
+        size_t templen = 0;
+        f->length = &templen;
+        ret = call_syscall(f, NULL);
+        if (ret != 0) {
+            return ret;
+        }
+    }
+    // Save current length, as syscall will update it
+    buflen = *f->length;
+    uint8_t *buf = malloc(buflen);
```
[the remark this answers is not in the document]
Comment (contrun) [reply lua-loader/lua-ckb.c:112]: I think I can set a maximum for the size of memory. But if the data is really that large, the lua script itself should pass a length argument.

File: lua-loader/lua-ckb.c (hunk 2 above)
Comment (contrun) [reply lua-loader/lua-ckb.c:94]: The offset argument is bundled in the struct syscall_function_t f. It is indeed passed.

File: lua-loader/lua-ckb.c (hunk 1 above)
Comment (XuJiandong) [reply lua-loader/lua-ckb.c:123]: The shrink operation is not necessary and wrong. What if the underlying data is very large (e.g. 300k) but we only fetch part of it (e.g. 1k)?

**Comment by XuJiandong** [comment]
Add a test case: witness has 500k data. Use load_witness to fetch part of it (e.g. 100 bytes).

**Pull request merged**